### PR TITLE
Backend user routes

### PR DESCRIPTION
**Pull request by jvendettuoli**
Apologies for the mess of a PR - at some point the `/dist` directory snuck into my git commits so deleting them takes up a ton of the files changed.

I rebuilt the user routes to get a feel for how NestJS works. They are now connected to our postgres db, and you can perform all CRUD operations on /api/users. I did not alter existing Auth routes, but will be doing so next to implement bcrypt encryption. Made minor change to test files to prevent failure, but they will eventually need a mock of TypeORM's import in the service files.

To test: 

- Create a user in db (doesn't require authentication) 
 ```
curl --request POST \
  --url http://localhost:3001/api/users \
  --header 'Content-Type: application/json' \
  --data '{
	"first_name": "TestFirst",
	"last_name": "TestLast",
	"email": "test@test.com",
	"password": "testpassword"
	}'
```
returns `{"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword","id":1}`

- Get users without auth token should fail 
```
curl --request GET \
  --url http://localhost:3001/api/users 
```
returns `{"statusCode":401,"message":"Unauthorized"}`

- Login to get access token
```
curl --request POST \
  --url http://localhost:3001/api/auth/login \
  --header 'Content-Type: application/json' \
  --data '{
	"username": "TestFirst",
	"password": "TestPass"
}'
```
returns an an object with an access_token property like `{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MjU5NDg1MTUsImV4cCI6MTYyNTk0ODU3NX0.VZb9X5Wkz0mxlN2hK32dcbKgP2h-BbQMZ39HPA0l2jA"}`

- Get users using the access_token should return all users. Replace the access token with the one you generated by /auth/login.
```
curl --request GET \
  --url http://localhost:3001/api/users \
  --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MjU5NDg2NTQsImV4cCI6MTYyNTk0ODcxNH0.xakTSjoaOay8eRLFhgsu9LztsXakeqamrPaCbB-rWYg'
```
returns all users. In my case `[{"id":1,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":2,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":3,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":4,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"}]`

Should be able to perform GET user by an id, PATCH user, and DELETE user as long as you have a valid token.

### CODE REVIEW

**Review by jd2rogers2 — APPROVED** [review]
sweet! couple question i left in there and one 👇 but def cool to login with the curl request

so i had to go into my pg db and alter cols of my user table (add email, pw)
if i didn't i would get
```
[Nest] 13784   - 07/12/2021, 6:58:51 AM   [TypeOrmModule] Unable to connect to the database. Retrying (1)... +744ms
QueryFailedError: column "email" of relation "users" contains null values
    at new QueryFailedError (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/typeorm/error/QueryFailedError.js:12:28)
    at PostgresQueryRunner.<anonymous> (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:248:31)
    at step (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:143:27)
    at Object.throw (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:124:57)
    at rejected (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:115:69)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

did you have to do this as well or is your docker container flexible and spun up based on entities?

File: server/src/users/users.service.ts [hunk 1]
```diff
@@ -1,34 +1,36 @@
...
     }
 
-    getByUsername(username: string): Promise<User> {
-        return this.userRepo.findOne({ first_name: username });
+    findOne(id: number) {
+        return this.usersRepository.findOne(id);
+    }
+    //Change to whatever the display name ends up being.
+    findByUsername(first_name: string) {
+        return this.usersRepository.findOne({first_name});
```
[the remark this answers is not in the document]
Comment (jvendettuoli) [reply server/src/users/users.service.ts:24]: You are correct, that is part of the bcrypt changes I made. I'll put that PR right after this.

File: server/src/database-connection.service.ts [hunk 2]
```diff
@@ -12,10 +12,10 @@ export class DatabaseConnectionService implements TypeOrmOptionsFactory {
             username: process.env.DATABASE_USER,
             password: process.env.DATABASE_PASSWORD,
             database: process.env.DATABASE_DB,
-            synchronize: true,
+            synchronize: true, //shouldn't be used in production
```
[the remark this answers is not in the document]
Comment (jvendettuoli) [reply server/src/database-connection.service.ts:15]: Yup! Sorry meant to do that once I looked into why that was the case. I still don't fully understand it, but it has to do with sync'ing the automatically loaded models, which I assume could result in a loss of data. I added the relevant link to the ticket.

File: server/src/users/entities/user.entity.ts [hunk 3]
```diff
@@ -1,16 +1,19 @@
 import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
```
[the remark this answers is not in the document]
Comment (jvendettuoli) [reply server/src/users/entities/user.entity.ts:1]: This was how I saw the NestJS Docs and a few tutorials structure the folder. It is also how it creates the resource when using `nest g resource`. I have no reason beyond be consistent with the docs I was learning from.

[the remark this answers is not in the document]
Comment (jvendettuoli) [reply server/src/users/entities/user.entity.ts:1]: `autoLoadEntities` applies for any entity passed into a module's imports like here 
```javascript
@Module({
  imports:[TypeOrmModule.forFeature([User])],
  controllers: [UsersController],
  providers: [UsersService],
  exports: [UsersService]
})
export class UsersModule {}
```
`autoLoadEntities` will load entities called by `TypeOrmModule.forFeature()`. As far as I know.

**Comment by jvendettuoli** [comment]
> sweet! couple question i left in there and one 👇 but def cool to login with the curl request
> 
> so i had to go into my pg db and alter cols of my user table (add email, pw)
> if i didn't i would get
> 
> ```
> [Nest] 13784   - 07/12/2021, 6:58:51 AM   [TypeOrmModule] Unable to connect to the database. Retrying (1)... +744ms
> QueryFailedError: column "email" of relation "users" contains null values
>     at new QueryFailedError (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/typeorm/error/QueryFailedError.js:12:28)
>     at PostgresQueryRunner.<anonymous> (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:248:31)
>     at step (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:143:27)
>     at Object.throw (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:124:57)
>     at rejected (/Users/jamesrogers/Dev/neh/web-app/server/node_modules/tslib/tslib.js:115:69)
>     at processTicksAndRejections (internal/process/task_queues.js:93:5)
> ```
> 
> did you have to do this as well or is your docker container flexible and spun up based on entities?


I did run into that error while getting it all up and running, but I thought it was related to not having it set up correctly. Upon reflection I think that the postgres docker container won't handle changes made automatically. It just spins up an empty container with what is needed to run postgres and the configuration we set. 

I believe what I had done that removed the error is this:

1. Within file `database-connection.service.ts` in the `/server` directory you'll find our database connection.
```javascript
@Injectable()
export class DatabaseConnectionService implements TypeOrmOptionsFactory {
    createTypeOrmOptions(): TypeOrmModuleOptions {
        return {
            name: 'default',
            type: 'postgres',
            host: process.env.DATABASE_HOST,
            port: Number(process.env.DATABASE_PORT),
            username: process.env.DATABASE_USER,
            password: process.env.DATABASE_PASSWORD,
            database: process.env.DATABASE_DB,
            synchronize: true, //shouldn't be used in production
            dropSchema: false, //toggle to true to clear database schema
            logging: true,
            autoLoadEntities: true 
        };
    }
}
```
2. `dropSchema` is how NestJS determines whether or not to handle dropping the existing schema (if it exists) and loading whatever is present when the server restarts.
3. Set `dropSchema:true`, restart the server, and the postgres database should be up to date (but also empty of any preexisting data).

I have it set to false normally so we can play around with persisting data between server restarts, but definitely set it to true and back to false a few time while figuring out how TypeORM and NestJS interacted with the postgres database. You are likely getting the error because the schema that exists now does not match that of your existing postgres database. My bad, I should have foreseen that and put it in the PR. In the future we'll need to drop the schema each time we change it. Also may need to delete the `dist` folder before server restart, because it will store the old schema in there I bet.

**Pull request merged**